File: 2024/haskell/solutions/Advent.hs
{-# LANGUAGE GADTs #-}

module Advent (Day, mkDay, exec, (|>)) where

import Data.Bifunctor (first)
import Data.Function ((&))
import Data.Text (Text)
import Data.Void (Void)
import Text.Megaparsec (Parsec, errorBundlePretty, runParser)

infixl 1 |>

(|>) :: a -> (a -> b) -> b
(|>) = (&)

exec :: Parsec Void Text a -> Text -> Either String a
exec parser = first errorBundlePretty . runParser parser ""

data Solution where
  Solution :: (Show a, Show b) => a -> b -> Solution

instance Show Solution where
  show (Solution a b) = "A: " <> show a <> "\nB: " <> show b

{- ORMOLU_DISABLE -}
type Parser a   = Text -> Either String a
type Solver a b = a    -> b
type Runner     = Text -> Either String Solution
type Day        = (Int, Runner)
{- ORMOLU_ENABLE -}

mkDay :: (Show b, Show c) => Int -> Parser a -> Solver a b -> Solver a c -> Day
mkDay day parse solveA solveB = (day, run)
  where
    run input = do
      text <- parse input
      let a = solveA text
      let b = solveB text
      pure $ Solution a b

File: 2024/rust/src/advent.rs
use std::{fmt::Debug, time::Instant};

pub type Error = Box<dyn std::error::Error>;

pub trait Solution: Debug {}

struct SolutionExists<A, B>(A, B);

impl<A, B> Debug for SolutionExists<A, B>
where
    A: Debug,
    B: Debug,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "A: {:?}", self.0)?;
        writeln!(f, "B: {:?}", self.1)?;
        Ok(())
    }
}

impl<A, B> Solution for SolutionExists<A, B>
where
    A: Debug,
    B: Debug,
{
}

type Parser<A> = fn(&str) -> Result<A, Error>;
type Solver<A, B> = fn(&A) -> B;
type Runner = Box<dyn Fn(&str) -> Result<Box<dyn Solution + 'static>, Error>>;
pub type Day = (usize, Runner);

pub fn mk_day<A, B, C>(day: usize, parse: Parser<A>, a: Solver<A, B>, b: Solver<A, C>) -> Day
where
    A: 'static,
    B: Debug + 'static,
    C: Debug + 'static,
{
    (
        day,
        Box::new(move |input| {
            let text = parse(input)?;

            let start = Instant::now();
            let a = a(&text);
            println!("Execution time A: {:?}", start.elapsed());

            let start = Instant::now();
            let b = b(&text);
            println!("Execution time B: {:?}", start.elapsed());

            Ok(Box::new(SolutionExists(a, b)))
        }),
    )
}

#[doc(inline)]
pub use crate::__day as day;

#[allow(clippy::crate_in_macro_def)]
#[doc(hidden)]
#[macro_export]
macro_rules! __day {
    ($day_num:expr, $parse:ident, $solve_a:ident, $solve_b:ident) => {
        pub fn day() -> crate::advent::Day {
            crate::advent::mk_day($day_num, $parse, $solve_a, $solve_b)
        }
    };
}
